[Docs] Updated Examples Section

aku1310 opened the pull request:
In **Other Examples**:
- Added more use cases of JSON Schema
- Added a short summary and index at the top

In **Miscellaneous Examples**:
- Removed Geographical Locations example as it is more suited in 'Other Examples' section
- Renamed it to Generic Examples

In **Modelling a File System**:
- Added a short summary before the index

Please let me know what you think about this and any changes required.

benjagm reviewed (changes requested):
Nice job. I am providing some small suggestions for better consistency among the different example pages.

We'll need more people reviewing the examples.

aku1310 commented:
> Nice job. I am providing some small suggestions for better consistency among the different example pages.
> 
> We'll need more people reviewing the examples.

Thank you! I've made the suggested changes.

benjagm reviewed (approved):
Looks good to me. Nice job!

I'd love to get at least another review.

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -41,7 +63,126 @@ title: JSON Schema examples
+
+## Blog post
+
+A schema representing a blog post, including properties like `title`, `content`, `published date`, `author details`, and `tags`.
```
If you're going to use the "speech" form of these properties (e.g. "published date" instead of `publishedDate`), I think it should be without formatting.  Otherwise, just use the actual property names.  Similarly for the other summaries.  We should be consistent as well.

spacether replied:
How about using the publishedDate form instead? Having a space in a key is not something that I have seen widely used in json in my experience.

gregsdennis replied:
Yes, that would fall under

> Otherwise, just use the actual property names.

gregsdennis reviewed (commented):
These examples look good, and it appears you put some time into them.  However, I would expect each example to showcase some different feature set.

For instance, if you're starting from the top, "blog post" adds nested properties, a feature that "address" doesn't have.  But then "book" just illustrates the same feature set that "address" did.  Does having "book," then, hold any real value?

To me, these all look like basically the same example.  (Maybe that's needed for this page; I'm not sure.)

aku1310 commented:
@gregsdennis @jeremyfiel Thank you for the reviews! I'll definitely modify the examples as I do see how they can seem repetitive and basic. Will research more on how I can improve them with your suggestions and then update the PR soon!

aku1310 commented:
I apologize for the delay in updating this PR 😅 

I have added some more properties to the examples such as `$dynamicRef`, `$anchor`, etc., and also removed some redundant examples which were showcasing something new.

Please let me know what you think about this.

gregsdennis commented on components/Sidebar.tsx:
```diff
@@ -39,7 +39,7 @@ const getDocsPath = [
   '/learn/json-schema-examples',
   '/learn/file-system',
-  '/learn/miscellaneous-examples',
+  '/learn/generic-examples',
```
I like miscellaneous here.  "Generic" already has specific meaning in programming, and we don't want to imply that these examples are that.

aku1310 replied:
I had actually updated the term after discussing with Benjamin on the slack thread ([here](https://json-schema.slack.com/archives/CB3SS5W5B/p1693215510648819?thread_ts=1688030108.211109&cid=CB3SS5W5B)), since miscellaneous and other seemed a bit similar as titles.😅 
I can change it back to Miscellaneous if that seems better, please let me know how to proceed with this.

benjagm replied:
> I can change it back to Miscellaneous

My bad @aku1310. @gregsdennis criteria has more weight here. Please let's rename it back to miscellaneous.

gregsdennis commented on pages/learn/generic-examples.md:
```diff
@@ -3,10 +3,9 @@ section: docs
 ---
 
-In this page, you will find examples illustrating different uses cases to help you get the most out of your JSON Schemas, including:
+In this page, you will find generic examples illustrating different uses cases to help you get the most out of your JSON Schemas. Each example comes with accompanying JSON data and explanation.
```
```suggestion
In this page, you will find miscellaneous examples illustrating different uses cases to help you get the most out of your JSON Schemas. Each example comes with accompanying JSON data and explanation.
```

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
+
+## Device type
+
+This schema represents electronic devices with a `deviceType` property that determines the device's category, such as `Smartphone` or `Laptop`.It employs the `$dynamicRef` keyword to dynamically reference schemas based on the `deviceType` property. This flexible schema structure allows data to conform to the appropriate device schema based on the `deviceType` specified, making it easy to describe ...
```
```suggestion
This schema represents electronic devices with a `deviceType` property that determines the device's category, such as `Smartphone` or `Laptop`. It employs the `$dynamicRef` keyword to dynamically reference schemas based on the `deviceType` property. This flexible schema structure allows data to conform to the appropriate device schema based on the `deviceType` specified, making it easy to describe different electronic devices with their unique characteristics
```

gregsdennis commented on pages/learn/json-schema-examples.md:
````diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
+```
+
+
+## Device type
````
I'm not sure what you're trying to do with this example.

The use of `$dynamicRef` is wrong,

- In 2020-12, `$dynamicRef` needs a `$dynamicAnchor` defined in the same schema resource, otherwise it works like an ordinary `$ref`.
- The value of `$dynamicRef` should be an anchor fragment (e.g. `#foo`), hence needing the anchor defined.
- The location `#/$defs/DeviceTypes` doesn't exist.

And I'm not sure a dynamic reference is the best solution for what the description says.  I'd use

- `oneOf` with `deviceType` defined using a `const` with a different value in each of the subschemas
- `if`/`then`/`else`
- follow [this example](https://json-schema.org/blog/posts/modelling-inheritance) to model inheritance (complex/advanced)

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
+        "orderId": { "type": "string" },
+        "items": {
+          "type": "array",
+          "items": { "$ref": "#/$defs/ProductSchema" }
```
If you're going to go to the trouble of defining an anchor, you should use it.

```suggestion
          "items": { "$ref": "#ProductSchema" }
```

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
+      }
+    },
+    "order": {
+      "$anchor": "OrderSchema",
```
This anchor isn't strictly necessary since it's not used anywhere, but leaving it isn't hurting anything.  (It might be used externally.)

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
+      "format": "date"
+    }
+  },
+  "$anchor": "JobPosting"
```
Having an anchor at the root isn't especially helpful.  You can do it, but why when

- internally you can just use `$ref: #` instead of `$ref: #JobPosting`
- externally, you can just `$ref: https://example.com/job-posting.schema.json` instead of `$ref: https://example.com/job-posting.schema.json#JobPosting`

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
     }
-  }
+  },
+  "$anchor": "MovieSchema"
```
Same comment as for `#JobPosting`

gregsdennis commented on pages/learn/json-schema-examples.md:
```diff
@@ -92,129 +177,403 @@ title: JSON Schema examples
     },
-    "nickname": {
+    "emergencyContact": {
+      "$ref": "https://example.com/user-profile.schema.json"
```
I like the idea of interlinking the examples 👍

gregsdennis reviewed (changes requested):
It looks like there are some misunderstandings around `$dynamic*` and `$anchor`.  It'd be good to review how those work.  Unfortunately we don't really have good docs for `$dynamic*` just yet (that I'm aware of), but here's [`$anchor`](https://json-schema.org/understanding-json-schema/structuring.html#anchor).

aku1310 commented:
> It looks like there are some misunderstandings around `$dynamic*` and `$anchor`. It'd be good to review how those work. Unfortunately we don't really have good docs for `$dynamic*` just yet (that I'm aware of) so you'll likely just need to try to read the spec (we can answer questions, too), but here's [`$anchor`](https://json-schema.org/understanding-json-schema/structuring.html#anchor).

Thank you! I'll go over this by the weekend, and update the PR soon.

aku1310 commented:
> Hey @aku1310, firsly, thanks again for taking the time and making the effort to try and improve our learning resources. Did you still want to continue to improve your PR here? If you're not wanting to spend much more time on it, maybe you could split some parts of it into a smaller PR, and then continue with other smaller PRs as time allows. Of course, please feel freeto ask further questions if you have any.

Hi, I apologise for the delay in this PR, I have not been able to properly work on this for a while 😅 Only the use cases in the other examples is left for now. Would it be alright to continue in this PR only? Or I can split it up if that is preferable. I aim to get this update the PR by this weekend as I have gotten free from other work now as well 😅

benjagm commented:
> Would it be alright to continue in this PR only?

That is totally fine Akanksha! We are very close to get this done. If you need help or feedback let me know and I'll be happy to help.

aku1310 commented:
> That is totally fine Akanksha! We are very close to get this done. If you need help or feedback let me know and I'll be happy to help.

Will surely ask, thank you!!

aku1310 commented:
> It looks like there are some misunderstandings around `$dynamic*` and `$anchor`. It'd be good to review how those work. Unfortunately we don't really have good docs for `$dynamic*` just yet (that I'm aware of) so you'll likely just need to try to read the spec (we can answer questions, too), but here's [`$anchor`](https://json-schema.org/understanding-json-schema/structuring.html#anchor).

Hi, as per the review, I have removed  `anchor` from `JobPosting` and `MovieSchema` as it was not helpful there, and also updated the `DeviceType` example using `oneOf` instead of `dynamicRef since `oneOf` does seems like a much better choice for that example. 

Please let me know what you think about this.

gregsdennis commented on pages/learn/json-schema-examples.md:
````diff
@@ -92,129 +177,414 @@ title: JSON Schema examples
+
+```json
+{
+  "deviceType": "Smartphone",
````
The casing here needs to match what the schema expects.  But the example looks a lot better now.

aku1310 replied:
Oh yes, I'll just update that 😅

gregsdennis reviewed (approved):
Overall much better.  Left one more comment; just a minor thing.

benjagm reviewed (approved):
Excellent job @aku1310!

Relequestual reviewed (approved):
Thanks for all your work on this. Very much appreciated!
